![](https://www.dii.uchile.cl/wp-content/uploads/2021/06/Magi%CC%81ster-en-Ciencia-de-Datos.png)

# Proyecto: Innovación Tecnológica en Bodoque Bank

**MDS7202: Laboratorio de Programación Científica para Ciencia de Datos**

### Cuerpo Docente:

- Profesor: Pablo Badilla, Ignacio Meza De La Jara
- Auxiliar: Sebastián Tinoco
- Ayudante: Diego Cortez M., Felipe Arias T.

*Por favor, lean detalladamente las instrucciones de la tarea antes de empezar a escribir.*


----

## Reglas

- Fecha de entrega: 01/06/2021
- **Grupos de 2 personas.**
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Estrictamente prohibida la copia. 
- Pueden usar cualquier material del curso que estimen conveniente.



# Proyecto

### Equipo:

- \<Raimundo Sandoval\>
- \<Cristóbal Subiabre\>


### Link de repositorio de GitHub: `https://github.com/raisando/Laboratorio-MDS7202->`




## 1. Introducción


El presente reporte tiene por finalidad entregar y evidenciar el procedimiento realizado para manipular las bases de datos entregadas. Buscando cargar los datos, realizar un EDA, limpiar los datos, detectar anormalias y levantar conclusiones pertinentes.

## 2. Lectura y Manejo de Datos




### 2.1 Cargar datos en dataframes

Primero se realiza la conexion con drive.

In [1]:
#!git clone https://gitlab.com/imezadelajara/datos_proyecto_1_mds7202.git

In [2]:
# Si usted está utilizando Colabolatory le puede ser útil este código para cargar los archivos.
try:
  from google.colab import drive
  import os
  drive.mount("/content/drive")
  #path = '/content/drive/MyDrive/U. Chile/9no Semestre 2023/Laboratorio Programación Cientifica/Proyecto_1'
  path = '/content/datos_proyecto_1_mds7202'
  os.chdir(path)
except: 
  print('Ignorando conexión drive-colab')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Luego, se procede a cargar los df

In [3]:
import pandas as pd

df_1 = pd.read_parquet("df_1.parquet")
df_2 = pd.read_parquet("df_2.parquet")
df_email_phone = pd.read_parquet("df_email_phone.parquet")

### 2.2 Exploración inicial

Ahora se revisan las primeras 5 filas de cada dataset.

In [4]:
display(df_1.head(5))

id  fraud_bool  income  prev_address_months_count  \
104446  729517           0     0.7                         -1   
269483  149585           0     0.8                         -1   
4102     64486           0     0.9                         -1   
351767  825283           0     0.5                         -1   
126377    8308           0     0.9                         -1   

        current_address_months_count  customer_age  days_since_request  \
104446                           305            60            0.030059   
269483                           140            50            0.015659   
4102                             171            50            0.001409   
351767                            85            30            0.027292   
126377                            39            30            0.010945   

        intended_balcon_amount payment_type  zip_count_4w  ...  \
104446               -1.599455           AC           990  ...   
269483                3.951994           AA          1269  ...   
4102                 28.159779           AB          4430  ...   
351767               -1.310498           AB          1698  ...   
126377               -1.450972           AC           569  ...   

        proposed_credit_limit  foreign_request    source  \
104446                  500.0                0  INTERNET   
269483                  200.0                0  INTERNET   
4102                   1500.0                0  INTERNET   
351767                  200.0                0  INTERNET   
126377                  200.0                0  INTERNET   

        session_length_in_minutes device_os  keep_alive_session  \
104446                   8.865992   windows                   0   
269483                   4.654872     linux                   1   
4102                     3.720953     linux                   1   
351767                   2.912670   windows                   1   
126377                   2.283680     other                   1   

       device_fraud_count  month        x1        x2  
104446                  0      2 -0.245425  0.568811  
269483                  0      7  0.009336 -2.096682  
4102                    0      1  2.229616 -0.005823  
351767                  0      3 -0.193945  0.861207  
126377                  0      3 -2.440650  0.354986  

[5 rows x 29 columns]

In [5]:
display(df_2.head(5))

id  fraud_bool  income  prev_address_months_count  \
699954  303612           0     0.8                         -1   
701206  568066           0     0.9                         -1   
270304  893344           0     0.7                         -1   
346538  247688           0     0.8                         -1   
147297  674478           0     0.8                         -1   

        current_address_months_count  customer_age  days_since_request  \
699954                            21            40            0.010837   
701206                            70            50            0.711485   
270304                           133            50            0.006856   
346538                           207            50            0.026194   
147297                            24            50            0.012646   

        intended_balcon_amount payment_type  zip_count_4w  ...  \
699954               -0.853976           AD           239  ...   
701206               25.137456           AA          1873  ...   
270304               -0.815330           AC           675  ...   
346538               -0.736615           AB           999  ...   
147297               10.744253           AA          3198  ...   

        proposed_credit_limit  foreign_request    source  \
699954                  200.0                0  INTERNET   
701206                  200.0                0  INTERNET   
270304                 1000.0                0  INTERNET   
346538                  200.0                0  INTERNET   
147297                  500.0                1  INTERNET   

        session_length_in_minutes device_os  keep_alive_session  \
699954                  48.127390     other                   0   
701206                   8.792273   windows                   1   
270304                  15.916225     other                   1   
346538                   0.509942     linux                   1   
147297                  10.562945     other                   0   

       device_fraud_count  month        x1        x2  
699954                  0      3  0.236931  0.077061  
701206                  0      3  0.571536  1.151563  
270304                  0      7  0.224252 -0.590459  
346538                  0      6  0.648528 -0.315911  
147297                  0      6  0.041475  0.383332  

[5 rows x 29 columns]

In [6]:
display(df_email_phone.head(5))

id  name_email_similarity  date_of_birth_distinct_emails_4w  \
987231  624588               0.555653                                15   
79954   620810               0.849718                                 2   
567130  580633               0.110898                                 3   
500891    6379               0.671250                                15   
55399   366511               0.772932                                 7   

        email_is_free  device_distinct_emails_8w  phone_home_valid  \
987231              0                          1                 0   
79954               0                          1                 0   
567130              0                          1                 1   
500891              0                          1                 0   
55399               0                          1                 0   

        phone_mobile_valid  
987231                   1  
79954                    0  
567130                   0  
500891                   1  
55399                    1

Además, se revisa el tipo de dato de cada columna usando info().

In [7]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 389782 entries, 104446 to 193209
Data columns (total 29 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   id                            389782 non-null  int64  
 1   fraud_bool                    389782 non-null  int64  
 2   income                        389782 non-null  float64
 3   prev_address_months_count     389782 non-null  int64  
 4   current_address_months_count  389782 non-null  int64  
 5   customer_age                  389782 non-null  int64  
 6   days_since_request            389782 non-null  float64
 7   intended_balcon_amount        389782 non-null  float64
 8   payment_type                  389782 non-null  object 
 9   zip_count_4w                  389782 non-null  int64  
 10  velocity_6h                   389782 non-null  float64
 11  velocity_24h                  389782 non-null  float64
 12  velocity_4w                   389782 no

In [8]:
df_2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 747410 entries, 699954 to 384611
Data columns (total 29 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   id                            747410 non-null  int64  
 1   fraud_bool                    747410 non-null  int64  
 2   income                        747410 non-null  float64
 3   prev_address_months_count     747410 non-null  int64  
 4   current_address_months_count  747410 non-null  int64  
 5   customer_age                  747410 non-null  int64  
 6   days_since_request            747410 non-null  float64
 7   intended_balcon_amount        747410 non-null  float64
 8   payment_type                  747410 non-null  object 
 9   zip_count_4w                  747410 non-null  int64  
 10  velocity_6h                   747410 non-null  float64
 11  velocity_24h                  747410 non-null  float64
 12  velocity_4w                   747410 no

In [9]:
df_email_phone.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 934730 entries, 987231 to 986114
Data columns (total 7 columns):
 #   Column                            Non-Null Count   Dtype  
---  ------                            --------------   -----  
 0   id                                934730 non-null  int64  
 1   name_email_similarity             934730 non-null  float64
 2   date_of_birth_distinct_emails_4w  934730 non-null  int64  
 3   email_is_free                     934730 non-null  int64  
 4   device_distinct_emails_8w         934730 non-null  int64  
 5   phone_home_valid                  934730 non-null  int64  
 6   phone_mobile_valid                934730 non-null  int64  
dtypes: float64(1), int64(6)
memory usage: 57.1 MB


De lo anterior se puede notar que los datos a priori se encuentran con el tipo de dato correcto y además, el df_1 tiene 389782 filas y 29 columnas, el df_2 tiene 747410 filas y 29 columans y por último, el df_email_phone tiene 934730 filas y 7 columnas.

### 2.3 Unir dataframes

Para unir los df de ejemplos se usa pd.concat

In [10]:
df_12 = pd.concat([df_1, df_2])

### 2.4 Unir nuevas variables

Se añaden las caracteristicas nuevas usando outer join segun "ID"

In [11]:
df = pd.merge(df_12, df_email_phone, on='id', how='outer')

### 2.5 Verificar nulos y duplicados, generar explicaciones sobre sus fuentes y proponer soluciones.



Se utilizan las funciones de pandas para revisar si hay nulos, eliminarlos y además revisar duplicados y eliminarlos.

In [12]:
display(df.isna().sum())

id                                      0
fraud_bool                          60959
income                              60959
prev_address_months_count           60959
current_address_months_count        60959
customer_age                        60959
days_since_request                  60959
intended_balcon_amount              60959
payment_type                        60959
zip_count_4w                        60959
velocity_6h                         60959
velocity_24h                        60959
velocity_4w                         60959
bank_branch_count_8w                60959
employment_status                   60959
credit_risk_score                   60959
housing_status                      60959
bank_months_count                   60959
has_other_cards                     60959
proposed_credit_limit               60959
foreign_request                     60959
source                              60959
session_length_in_minutes           60959
device_os                         

Se observa que existe una gran cantidad de filas vacias, ello puede ser debido a un fallo en la base de datos y no haya sido capaz de obtener los valores para dichos valores. Por otro lado, se ve que las últimas columnas tienen más NAN, ello se puede deber a que hay usuarios los cuales no tenian datos extras en el tercer dataset. Por ende, para regularizar la situacion se van a eliminan dichas filas.

In [13]:
df = df.dropna()

Se revisan la existencia de filas duplicadas.

In [14]:
print(df.duplicated().sum())

183694


Se ve que hay varios valores duplicados, lo cual podria deberse a clientes duplicados en ambas datos datasets (df_1, df_2). De modo que se eliminan los duplicados

In [15]:
df = df.drop_duplicates()

### 2.6 Limpiar

Se eliminan las filas con NAN y Duplicadas

In [16]:
df = df.dropna()
df = df.drop_duplicates()

Con ello el dataset resultante tiene una dimension de:

In [17]:
df.shape

(873771, 35)

### 2.7 Segmentación Etaria
Para ello se asigna segun el siguiente numero el rango etario

1. Menor de 18 años.
2. Entre 18 y 26 años.
3. Entre 27 y 59 años.
4. Mayor a 60 años.

In [18]:
def segmentacion_etaria(edad):
    if edad < 18:
        return 1
    elif edad >= 18 and edad <= 26:
        return 2
    elif edad >= 27 and edad <= 59:
        return 3
    else:
        return 4

In [19]:
df['segmentacion_etaria'] = df['customer_age'].apply(segmentacion_etaria)

## 3. Análisis Exploratorio de Datos

In [20]:
#!pip install ydata_profiling

In [21]:
from ydata_profiling import ProfileReport

#profile = ProfileReport(df, title="EDA")
#profile.to_file("eda.html")

In [22]:
#profile

### 3.1 Análisis del EDA

En primer lugar, vemos que se tiene un 0% de datos nulos y duplicados.

Podemos confirmar rapidamente tambien que la cantidad de instancias etiquetadas como fraude son notablemente inferiores a las no-fraude.

En cuanto a las distribuciones, es notable que existen variables con una distribucion psuedo-exponencial, como:

*  prev_address_months_count
*  current_address_months_count
*  days_since_request
*  intended_balcon_amount
*  bank_branch_count_8w

Esta ultima muestra una clara concentracion de valores 0, pero no a niveles preocupantes que demuestren una falla en la data, si no mas bien un comportamiento natural de los clientes del banco.

Vemos en segundo lugar un numero de variables con una distribucion normal skewed a la izquierda, o tipo poisson, como:
*   date_of_birth_distinct_emails_4w
*   velocity_6h

Aqui tambien vemos como la *customer_age* que esta levemente desplazada a la izquierda, tiene un peak particular en los 50 años, el *zip_count_4w*, que tiene una segunda subida al rededor del 2500, o el *session_length_in_minutes* que se ve muy concentrado en valores menores a 10.

El *credit_risk_score* pareciera ser la unica feature con una distribucion  absolutamente normal, esto probablemente debido a que justamente es una variable construida sobre los datos. *Velocity_24h* pareciera aproximarse muy de cerca a una normal, pero contiene una especie de segundo peak aproximadamente a en el valor 3000. 

Algunas otras variables muestran una distribucion desconocida, como el:
*   income
*   velocity_4w
*   bank_months_count
*   proposed_credit_limit
*   name_email_similarity

En cuanto a outliers, solo *name_email_similarity* y *intended_balcon_amount* muestran señales un poco mas claras de presencia de outliers, puesto que sus ditribuciones muestran una tendencia clara antes de los valores extremos, mientras que otras como *bank_branch_count_8w*, o *proposed_credit_limit* que tambien tienen valores extremos, parecen seguir un patron mas o menos comun, tal que no podemos atribuir estas "lomas" en valores extremos tan concluyentemente a errores en la data.

Se ve tambien un claro desbalance dentro de las variables categoricas como:
*   source
*   foreign_request
*   has_other_cards

Finalmente, la variable *device_fraud_count* contiene solo valores 0.

Continuando con el analisis, *customer_age* y *segmentacion_etaria* estan evidentemente altamente correlacionadas. Pero vemos tambien que *current_address_months_count* y *prev_address_months_count* tienen un alto de valor de correlacion (pero inverso), lo mismo para *month* y las *velocity*s.

## 4. Preprocesamiento

### 4.1 Declarar `ColumnTransformer`

Se define los preprocesadores para cada tipo de dato y posteriormente se crea un ColumnTransformer de la libreria scikit learn para realizar el procesamiento, asignando cada columna a la categoria de categorica u ordinal segun corresponde, en conjunto con el preprocesador, además, dado a que id y fraud_bool no deben ser transformadas pero si mantenidas en el df se utiliza el termino passthrough.

In [24]:
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder
from sklearn.compose import ColumnTransformer

#Se definen los preprocesadores para cada tipo de dato
preprocesador_categorico = OneHotEncoder()
preprocesador_ordinal = OrdinalEncoder()

# ColumnTransformer para aplicar transformaciones a las columnas
preprocesamiento = ColumnTransformer(
    transformers=[
        ('categorical', preprocesador_categorico, ['payment_type', 'employment_status', 'housing_status', 'has_other_cards', 'foreign_request', 'source',  'device_os', 'keep_alive_session',  'device_fraud_count',  'email_is_free', 'device_distinct_emails_8w', 'phone_home_valid', 'phone_mobile_valid', 'segmentacion_etaria']),
        ('ordinal', preprocesador_ordinal, ['income', 'prev_address_months_count', 'current_address_months_count', 'customer_age', 'days_since_request', 'intended_balcon_amount', 'zip_count_4w', 'velocity_6h', 'velocity_24h', 'velocity_4w', 'bank_branch_count_8w', 'credit_risk_score', 'bank_months_count', 'proposed_credit_limit', 'session_length_in_minutes', 'month', 'x1', 'x2', 'name_email_similarity', 'date_of_birth_distinct_emails_4w']),
    ],
    remainder='passthrough'  # Mantener las columnas no transformadas (id y fraud_bool)
)

### 4.2 Transformar datos

Se aplica la transfromación de datos anteriormente definida y se almacena en un df, el cual posteriormente se le modifica el nombre de las columnas segun los datos anteriores.

In [25]:
df_original = df.copy()
df = pd.DataFrame(preprocesamiento.fit_transform(df_original))

In [26]:
df.columns = preprocesamiento.get_feature_names_out(df_original.columns)

## 5. Visualización en Baja Dimensionalidad

### 5.1 Muestrear dataframe

Se muestrean los datos transformados anteriormente y se elimina la columna 'fraud_bool' para evitar el data leakage.

In [27]:
df.head(5)

categorical__payment_type_AA  categorical__payment_type_AB  \
0                           0.0                           0.0   
1                           1.0                           0.0   
2                           0.0                           1.0   
3                           0.0                           1.0   
4                           0.0                           0.0   

   categorical__payment_type_AC  categorical__payment_type_AD  \
0                           1.0                           0.0   
1                           0.0                           0.0   
2                           0.0                           0.0   
3                           0.0                           0.0   
4                           1.0                           0.0   

   categorical__payment_type_AE  categorical__employment_status_CA  \
0                           0.0                                0.0   
1                           0.0                                0.0   
2                           0.0                                0.0   
3                           0.0                                1.0   
4                           0.0                                1.0   

   categorical__employment_status_CB  categorical__employment_status_CC  \
0                                0.0                                1.0   
1                                0.0                                0.0   
2                                0.0                                1.0   
3                                0.0                                0.0   
4                                0.0                                0.0   

   categorical__employment_status_CD  categorical__employment_status_CE  ...  \
0                                0.0                                0.0  ...   
1                                0.0                                0.0  ...   
2                                0.0                                0.0  ...   
3                                0.0                                0.0  ...   
4                                0.0                                0.0  ...   

   ordinal__bank_months_count  ordinal__proposed_credit_limit  \
0                         0.0                             4.0   
1                         9.0                             1.0   
2                         1.0                             8.0   
3                         3.0                             1.0   
4                         0.0                             1.0   

   ordinal__session_length_in_minutes  ordinal__month  ordinal__x1  \
0                            633893.0             2.0     350513.0   
1                            374377.0             7.0     437783.0   
2                            272416.0             1.0     859970.0   
3                            195728.0             3.0     367922.0   
4                            145086.0             3.0       6356.0   

   ordinal__x2  ordinal__name_email_similarity  \
0     622855.0                        786689.0   
1      15572.0                         98353.0   
2     432958.0                        688171.0   
3     701414.0                        758198.0   
4     555761.0                        404812.0   

   ordinal__date_of_birth_distinct_emails_4w  remainder__id  \
0                                        2.0       729517.0   
1                                        4.0       149585.0   
2                                        4.0        64486.0   
3                                       10.0       825283.0   
4                                       13.0         8308.0   

   remainder__fraud_bool  
0                    0.0  
1                    0.0  
2                    0.0  
3                    0.0  
4                    0.0  

[5 rows x 69 columns]

In [28]:
df_muestreado = df.sample(n=20000, random_state=1999)
features = df_muestreado.drop('remainder__fraud_bool', axis=1)

### 5.2 Proyectar y agregar proyecciones al dataframe de muestreo

Primero se instala la libreria de umap-learn

In [30]:
#!pip install umap-learn

Luego se escalan los datos mediante minmax scaler y se realiza la proyección con umap, esta proyección se agrega al dataset muestreado.

In [31]:
import umap
from sklearn.preprocessing import MinMaxScaler

scaled_features = MinMaxScaler().fit_transform(features)

umap = umap.UMAP()

# Realizar la proyección
df_umap_embedded = umap.fit_transform(scaled_features)

#Se agrega la proyección al df transformado
df_muestreado["x_umap"] = df_umap_embedded[:, 0]
df_muestreado["y_umap"] = df_umap_embedded[:, 1]

### 5.3 Visualizar según rangos etarios

Primero se realiza la segmentación etaria de acuerdo a lo que se pide en el enunciado.

In [32]:
df_muestreado_se_1 = df_muestreado[df_muestreado['categorical__segmentacion_etaria_1'] == 1]
df_muestreado_se_2 = df_muestreado[df_muestreado['categorical__segmentacion_etaria_2'] == 1]
df_muestreado_se_3 = df_muestreado[df_muestreado['categorical__segmentacion_etaria_3'] == 1]
df_muestreado_se_4 = df_muestreado[df_muestreado['categorical__segmentacion_etaria_4'] == 1]

Ahora se generan las graficas para cada rango etario y se colorean segun la variable 'fraud_bool'

In [33]:
import plotly.express as px
from plotly.subplots import make_subplots

fig_se_1 = px.scatter(df_muestreado_se_1, x='x_umap', y='y_umap', color='remainder__fraud_bool', color_continuous_scale='Viridis' , title='UMAP 0-17')
fig_se_2 = px.scatter(df_muestreado_se_2, x='x_umap', y='y_umap', color='remainder__fraud_bool',  color_continuous_scale='Viridis' ,title='UMAP 18-26')
fig_se_3 = px.scatter(df_muestreado_se_3, x='x_umap', y='y_umap', color='remainder__fraud_bool',  color_continuous_scale='Viridis',title='UMAP 27-59')
fig_se_4 = px.scatter(df_muestreado_se_4, x='x_umap', y='y_umap', color='remainder__fraud_bool',  color_continuous_scale='Viridis',title='UMAP 60+')

fig = (
    make_subplots(rows=2, cols=2)
    .add_trace(fig_se_1.data[0], row=1, col=1)
    .add_trace(fig_se_2.data[0], row=1, col=2)
    .add_trace(fig_se_3.data[0], row=2, col=1)
    .add_trace(fig_se_4.data[0], row=2, col=2)

    # Añadir títulos a cada gráfica
    .update_xaxes(title_text="UMAP 0-17", row=1, col=1)
    .update_xaxes(title_text="UMAP 18-26", row=1, col=2)
    .update_xaxes(title_text="UMAP 27-59", row=2, col=1)
    .update_xaxes(title_text="UMAP 60+", row=2, col=2)

    .update_layout(
        height=800,
        title_text="Comparación UMAP x Segmento etario",
        legend=dict(yanchor="top", y=0.99, xanchor="left", x=0.01)
    )
    .show()
)

Cabe destacar que a partir de las graficas se puede ver que a pesar de la diferencia de tamaño de cada muestra, estas se distribuyen de manera similar tanto cuando fraud_bool = 0 y cuando fraud_bool = 1, luego en cuanto a la densidad de cada una de las graficas, se ve que se debe al desbalance presente en el dataset. 

## 6. Anomalías

### 6.1 Implementar detector de anomalías sobre dataframe de muestreo

Se aplica el metodo de Isolation Forest para detectar anomalias, individualmente para cada segmento etario.

In [76]:
import numpy as np
from sklearn.ensemble import IsolationForest

data1 = df_muestreado_se_1.copy()
y_se1 = data1.remainder__fraud_bool
X_se1 = data1.drop(columns=['remainder__fraud_bool','remainder__id'])
scaled_features_se1 = MinMaxScaler().fit_transform(X_se1)

isf = IsolationForest(n_estimators=100)
outliers_se1 = isf.fit_predict(scaled_features_se1)
#px.scatter(data, x="x_umap", y="y_umap", color=outliers.astype(str))

In [77]:
data2 = df_muestreado_se_2.copy()
y_se2 = data2.remainder__fraud_bool
X_se2 = data2.drop(columns=['remainder__fraud_bool','remainder__id'])
scaled_features_se2 = MinMaxScaler().fit_transform(X_se2)

isf = IsolationForest(n_estimators=100)
outliers_se2 = isf.fit_predict(scaled_features_se2)
#px.scatter(data, x="x_umap", y="y_umap", color=outliers.astype(str))

In [78]:
data3 = df_muestreado_se_3.copy()
y_se3 = data3.remainder__fraud_bool
X_se3 = data3.drop(columns=['remainder__fraud_bool','remainder__id'])
scaled_features_se3 = MinMaxScaler().fit_transform(X_se3)

isf = IsolationForest(n_estimators=100)
outliers_se3 = isf.fit_predict(scaled_features_se3)
#px.scatter(data, x="x_umap", y="y_umap", color=outliers.astype(str))

In [79]:
data4 = df_muestreado_se_4.copy()
y_se4 = data4.remainder__fraud_bool
X_se4 = data4.drop(columns=['remainder__fraud_bool','remainder__id'])
scaled_features_se4 = MinMaxScaler().fit_transform(X_se4)

isf = IsolationForest(n_estimators=100)
outliers_se4 = isf.fit_predict(scaled_features_se4)
#px.scatter(data, x="x_umap", y="y_umap", color=outliers.astype(str))

### 6.2 Agregar resultados a dataframe de muestreo

In [80]:
df_muestreado_se_1['outlier'] = outliers_se1
df_muestreado_se_2['outlier'] = outliers_se2
df_muestreado_se_3['outlier'] = outliers_se3
df_muestreado_se_4['outlier'] = outliers_se4

df_muestreado_reconst = pd.concat([df_muestreado_se_1,df_muestreado_se_2,df_muestreado_se_3,df_muestreado_se_4])

<ipython-input-80-adfb6a0802a4>:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

<ipython-input-80-adfb6a0802a4>:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

<ipython-input-80-adfb6a0802a4>:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

<ipython-input-80-adfb6a0

### 6.3 Visualizar según rangos etarios

In [81]:
fig_se_1 = px.scatter(df_muestreado_se_1, x="x_umap", y="y_umap", color=df_muestreado_se_1.outlier, title='UMAP 0-17')
fig_se_2 = px.scatter(df_muestreado_se_2, x="x_umap", y="y_umap", color=df_muestreado_se_2.outlier,title='UMAP 18-26')
fig_se_3 = px.scatter(df_muestreado_se_3, x="x_umap", y="y_umap", color=df_muestreado_se_3.outlier,title='UMAP 27-59')
fig_se_4 = px.scatter(df_muestreado_se_4, x="x_umap", y="y_umap", color=df_muestreado_se_4.outlier,title='UMAP 60+')

fig = (
    make_subplots(rows=2, cols=2)
    .add_trace(fig_se_1.data[0], row=1, col=1)
    .add_trace(fig_se_2.data[0], row=1, col=2)
    .add_trace(fig_se_3.data[0], row=2, col=1)
    .add_trace(fig_se_4.data[0], row=2, col=2)

    # Añadir títulos a cada gráfica
    .update_xaxes(title_text="UMAP 0-17", row=1, col=1)
    .update_xaxes(title_text="UMAP 18-26", row=1, col=2)
    .update_xaxes(title_text="UMAP 27-59", row=2, col=1)
    .update_xaxes(title_text="UMAP 60+", row=2, col=2)

    .update_layout(
        height=800,
        title_text="Comparación UMAP x Segmento etario",
        legend=dict(yanchor="top", y=0.99, xanchor="left", x=0.01)
    )
    .show()
)

### 6.4 Calcular ratios y responder

In [86]:
pred_correcto = len(df_muestreado_reconst[(df_muestreado_reconst['remainder__fraud_bool'] == 1) & (df_muestreado_reconst['outlier'] == -1)])

pred_fraude = len(df_muestreado_reconst[df_muestreado_reconst['outlier'] == -1])

P = pred_correcto / pred_fraude

real_fraude = len(df_muestreado_reconst[df_muestreado_reconst['remainder__fraud_bool'] == 1])

R = pred_correcto / real_fraude


In [87]:
print('Ratio P |Cantidad de ejemplos predichos correctamente como fraude / cantidad total de datos predichos como fraude: ', P)
print('Ratio R |Cantidad de ejemplos predichos correctamente como fraude / cantidad total de ejemplos que eran realmente fraude: ', R)


Ratio P |Cantidad de ejemplos predichos correctamente como fraude / cantidad total de datos predichos como fraude:  0.012428454619787407
Ratio R |Cantidad de ejemplos predichos correctamente como fraude / cantidad total de ejemplos que eran realmente fraude:  0.3568075117370892


Estos ratios representan las metricas de precision y recall, respectivamente se pueden interpretar como la cantidad de verdaderos positivos, y la cobertura de el modelo. Es decir si tengo precision 0.90 y recall 0.85, significa que tengo un 10% de falsos positivios, y que "atrapo" al 85% de los fraudulentos.

En este caso podemos ver que tenemos una pesima performance para encontrar a los datos fraudulentos, inferiores incluso al 10%. Estos son coherentes con las visualizaciones mostradas anteriormente, donde se ve claramente que el metodo marca muchos mas casos como fraude, y estos no tienen relamente ninguna cercania visualmente a los verdaderos fraudulentos que vemos mas arriba. 

Esto se puede deber a que los casos de fraude no tienen ningun comportamiento comun, ni entre grupos etarios ni en general. Incluso podemos ver que aplicando la tecnica de UMAP, estos se distribuyen "dentro" de la misma nube de casos no fraudulentos, y no son distinguibles ni separables facilmente mediante este tipo de tecnicas de clustering o deteccion de outliers. 